accepted/2020/linking-libraries.md

```diff
@@ -0,0 +1,152 @@
+
+Since these 2 scenarios require the use of Embedded Descriptor xml files, we will need to teach the linker to respect feature switches inside an Embedded Descriptor xml file.
+
+#### Analyzing Canonical Applications
+
+Application size is similar to throughput performance in that you can't guess at what could effect an application size. You have to measure it. In order to measure it, we will need to define a set of canonical applications that we want to measure. Once we have these applications, we can link them and then analyze the output to see what pieces are remaining, why, and investigate how to remove the unnecessary pieces.
+
+ILLink is capable of saving the dependency graph when it trims an application. There is a tool, [Linker Analyzer](https://github.com/mono/linker/tree/d3eb4ea8179516b0554a3d3089c71062496bd6f0/src/analyzer), that can load this graph and allows a developer to analyze it. Example questions that can be answered are:
+
+1. What are the biggest types/methods left after trimming?
+2. What is the call chain that is keeping type/method `foo` from being trimmed?
+
+Today the Linker Analyzer is a console application. One enhancement we can make is to put the core logic into a NuGet package, and allow a developer to analyze their application using the [.NET Interactive](https://github.com/dotnet/interactive/blob/master/README.md) notebook experience. Advantages of this experience are:
+
+1. The dependency graph can be large (many MBs). Loading and unloading it to ask one question at a time in a console app takes time. In a notebook, it is loaded into memory once and cached. Many questions can be asked faster.
+2. Developers can write their own logic and extensions to the dependency graph.
+3. We can have template notebooks that show examples of how to analyze an application's size.
+4. When a new feature is added to the Analyzer, we just need to create a public API. We don't need to create command line arguments.
+5. We have the option of visuals and more interactive output - for example hyperlinks. In a console application, you are stuck with text.
+
+## Testing
+
+Testing the linked applications falls into two buckets:
+1. Does the application still work?
+2. Is the application smaller?
+
+In general, using unit tests to ensure a linked application still works isn't a great testing strategy. Unit tests, by definition, try exercising all the code in the library. If all the code in the library is being exercised, the linker won't be able to trim any of it. Thus if nothing is trimmed, you aren't confirming that the library still works when not all of the code is present.
+
+Instead of using unit tests, we should create dedicated applications to test that the libraries are linker safe. One set of applications will be higher-level applications that we want to measure the size over time. Another set will target specific APIs that needed to be annotated to ensure the annotations and the linker are working correctly.
```

Comment:
Could this be rewritten to be more actionable? 
	- How should target specific APIs test look like?
	- Where they will be located?
	- How they will be run?
	- Who will be responsible for writing the test(s)?

Reply:
> Who will be responsible for writing the test(s)?

The .NET Library developers will write the target specific APIs test, as it is API specific.

As for the higher-level applications, I think that is up for debate whether it is the perf team, the application-model team (Xamarin, Blazor, etc), the Libraries team, or someone else.

> Where they will be located?

I think a decent place for the higher-level applications would be https://github.com/dotnet/performance. It has the ability to do the "measure over time" capabilities that I mention in the doc. It would also keep application-model specific dependencies out of dotnet/runtime.

As for the targeted API tests, I was assuming they would be in dotnet/runtime, since they are tied to the APIs.

Reply:
I'd focus on unit test only in this doc but define them well (see the questions I asked)